File: swift/Sources/MoneyHero/MarketData/MarketQuoteFetcher.swift
struct MarketQuoteFetchResult: Sendable {
    let symbol: String
    let quote: MarketQuote
    let history: HistoricalPriceSeries?
}

struct MarketQuoteFetcher: Sendable {
    private let provider: any MarketDataProviding
    private let parser: any MarketChartParsing
    private let goldSymbol = YahooFinanceClient.goldSymbol

    init(provider: any MarketDataProviding, parser: any MarketChartParsing) {
        self.provider = provider
        self.parser = parser
    }

    func fetch(symbol rawSymbol: String, kind: MarketRefreshTrackedSymbolKind) async throws -> MarketQuoteFetchResult {
        let symbol = try normalizeSymbol(rawSymbol)

        switch kind {
        case .currency:
            return try await fetchCurrency(symbol: symbol)
        case .market where symbol == goldSymbol:
            return try await fetchGold(symbol: symbol)
        case .market:
            return try await fetchStock(symbol: symbol)
        }
    }

    private func fetchCurrency(symbol: String) async throws -> MarketQuoteFetchResult {
        if symbol == "USD" {
            return try await fetchUSDIndex()
        }

        let response = try await provider.fetchCurrencyChart(displayedCurrency: symbol)
        let quote = try parser.parseCurrencyQuote(
            data: response.data,
            requestedSymbol: symbol,
            displayedCurrency: symbol,
            invertForUSDDisplay: true
        )
        return MarketQuoteFetchResult(symbol: symbol, quote: quote, history: nil)
    }

    private func fetchUSDIndex() async throws -> MarketQuoteFetchResult {
        let indexSymbol = YahooFinanceClient.usDollarIndexSymbol
        let response = try await provider.fetchStockChart(symbol: indexSymbol)
        let parsed = try parser.parseStockQuoteAndHistory(data: response.data, requestedSymbol: indexSymbol)
        let quote = fixedUSDQuote(from: parsed.quote, indexSymbol: indexSymbol)
        return MarketQuoteFetchResult(symbol: "USD", quote: quote, history: nil)
    }

    private func fetchGold(symbol: String) async throws -> MarketQuoteFetchResult {
        let quoteResponse = try await provider.fetchGoldQuoteChart()
        let historyResponse = try await provider.fetchGoldHistoryChart()
        let parsed = try parser.parseGoldQuoteAndHistory(
            quoteData: quoteResponse.data,
            historyData: historyResponse.data
        )
        return MarketQuoteFetchResult(symbol: symbol, quote: parsed.quote, history: parsed.history)
    }

    private func fetchStock(symbol: String) async throws -> MarketQuoteFetchResult {
        let response = try await provider.fetchStockChart(symbol: symbol)
        let parsed = try parser.parseStockQuoteAndHistory(data: response.data, requestedSymbol: symbol)
        return MarketQuoteFetchResult(symbol: symbol, quote: parsed.quote, history: parsed.history)
    }

    private func fixedUSDQuote(from indexQuote: MarketQuote, indexSymbol: String) -> MarketQuote {
        let previousClose = fixedPreviousClose(changePercent: indexQuote.changePercent)
        return MarketQuote(
            symbol: "USD",
            providerSymbol: indexQuote.providerSymbol ?? indexSymbol,
            name: "US Dollar Index",
            provider: indexQuote.provider,
            price: 1,
            previousClose: previousClose,
            change: 1 - previousClose,
            changePercent: indexQuote.changePercent,
            currency: "USD",
            marketTime: indexQuote.marketTime,
            exchangeTimezoneName: indexQuote.exchangeTimezoneName,
            change1m: indexQuote.change1m,
            change6m: indexQuote.change6m,
            change1y: indexQuote.change1y,
            change2y: indexQuote.change2y,
            dividends: []
        )
    }

    private func fixedPreviousClose(changePercent: Double) -> Double {
        let denominator = 1 + (changePercent / 100)
        guard denominator.isFinite, denominator > 0 else {
            return 1
        }

        return 1 / denominator
    }
}

File: flutter/lib/market_data/market_quote_fetcher.dart
import "../domain/market_history_models.dart";
import "../domain/market_normalization.dart";
import "../domain/market_provider_models.dart";
import "../domain/market_quote_models.dart";
import "market_data_protocols.dart";
import "market_refresh_request_planner.dart";
import "yahoo_finance_client.dart";

class MarketQuoteFetchResult {
  const MarketQuoteFetchResult({required this.symbol, required this.quote, required this.history});

  final String symbol;
  final MarketQuote quote;
  final HistoricalPriceSeries? history;
}

class MarketQuoteFetcher {
  const MarketQuoteFetcher({required this.provider, required this.parser});

  final MarketDataProviding provider;
  final MarketChartParser parser;

  Future<MarketQuoteFetchResult> fetch({required String symbol, required MarketRefreshTrackedSymbolKind kind}) async {
    final String normalized = normalizeSymbol(symbol);
    if (kind == MarketRefreshTrackedSymbolKind.currency) {
      if (normalized == "USD") {
        return const MarketQuoteFetchResult(
          symbol: "USD",
          quote: MarketQuote(symbol: "USD", providerSymbol: null, name: "US Dollar", provider: MarketDataProvider.yahoo, price: 1, previousClose: 1, change: 0, changePercent: 0, currency: "USD", marketTime: null, exchangeTimezoneName: null, change1m: null, change6m: null, change1y: null, change2y: null, dividends: <DividendEvent>[]),
          history: null,
        );
      }
      final response = await provider.fetchCurrencyChart(displayedCurrency: normalized);
      final quote = parser.parseCurrencyQuote(
        data: response.data,
        requestedSymbol: normalized,
        displayedCurrency: normalized,
        invertForUsdDisplay: true,
      );
      return MarketQuoteFetchResult(symbol: normalized, quote: quote, history: null);
    }
    if (normalized == YahooFinanceClient.goldSymbol) {
      final quoteResponse = await provider.fetchGoldQuoteChart();
      final historyResponse = await provider.fetchGoldHistoryChart();
      final parsed = parser.parseGoldQuoteAndHistory(quoteData: quoteResponse.data, historyData: historyResponse.data);
      return MarketQuoteFetchResult(symbol: normalized, quote: parsed.quote, history: parsed.history);
    }
    final response = await provider.fetchStockChart(symbol: normalized);
    final parsed = parser.parseStockQuoteAndHistory(data: response.data, requestedSymbol: normalized);
    return MarketQuoteFetchResult(symbol: normalized, quote: parsed.quote, history: parsed.history);
  }

}
